pkgs/by-name/rk/rkboot/package.nix

```diff
@@ -0,0 +1,38 @@
+{ stdenv
+, lib
+, rkbin
+, box64
+}:
+
+stdenv.mkDerivation {
+  name = "rkboot";
+
+  src = rkbin.src;
+
+  postPatch = ''
+    substituteInPlace RKBOOT/*.ini --replace 'PATH=' 'PATH=rkboot/'
+  '';
+
+  buildPhase = ''
+    mkdir rkboot
+    for i in $(ls ./RKBOOT/*.ini)
+    do
+      # The proprietary, statically linked binaries to perform boot_merge are
+      # x86_64 only. Though we use box64 to emulate if building on aarch64-linux
+      ${lib.optionalString stdenv.isAarch64 "${lib.getExe box64}"} ./tools/boot_merger "$i" || true
+    done
+  '';
+
+  installPhase = ''
+    mkdir -p $out
+    mv rkboot $out/bin
```

Comment:
When running on aarch64, this only produces an empty directory:

```
sh: lscpu: not found
Running on Unknown CPU with 8 Cores
Params database has 14 entries
Box64 with Dynarec v0.2.6 nogit built on Jan  1 1980 00:00:00
BOX64: Detected 48bits at least of address space
Error: PageSize configuration is wrong: configured with 4096, but got 16384
Dynarec for ARM64, with extension: ASIMD AES CRC32 PMULL SHA1 SHA2 USCAT FLAGM FLAGM2 FRINT PageSize:16384 sh: lscpu: not found
sh: lscpu: not found
Running on Unknown CPU with 8 Cores
Params database has 14 entries
Box64 with Dynarec v0.2.6 nogit built on Jan  1 1980 00:00:00
BOX64: Detected 48bits at least of address space
Error: PageSize configuration is wrong: configured with 4096, but got 16384
Dynarec for ARM64, with extension: ASIMD AES CRC32 PMULL SHA1 SHA2 USCAT FLAGM FLAGM2 FRINT PageSize:16384 sh: lscpu: not found
sh: lscpu: not found
Running on Unknown CPU with 8 Cores
Params database has 14 entries
Box64 with Dynarec v0.2.6 nogit built on Jan  1 1980 00:00:00
BOX64: Detected 48bits at least of address space
Error: PageSize configuration is wrong: configured with 4096, but got 16384
Running phase: installPhase
Running phase: fixupPhase
shrinking RPATHs of ELF executables and libraries in /nix/store/1aliw7sava4z2yijmh9vx9f77qi323qb-rkboot
checking for references to /build/ in /nix/store/1aliw7sava4z2yijmh9vx9f77qi323qb-rkboot...
patching script interpreter paths in /nix/store/1aliw7sava4z2yijmh9vx9f77qi323qb-rkboot
stripping (with command strip and flags -S -p) in  /nix/store/1aliw7sava4z2yijmh9vx9f77qi323qb-rkboot/bin
/nix/store/1aliw7sava4z2yijmh9vx9f77qi323qb-rkboot
 flokli@m2air  ~/nixpkgs   master ↑39600  find result/
result/
result/bin
```

Can you fix this, or make it x86_64-linux only?

Reply:
@flokli I think you may be running box64 with binfmt on aarch64. So the empty directory is caused by an impurity and error in your emulator? If you run this on a real aarch64 machine, the directory is not empty.

```
[matthewcroughan@aarch64:~]$ uname -a
Linux aarch64.nixos.community 5.15.156 #1-NixOS SMP Wed Apr 17 09:15:18 UTC 2024 aarch64 GNU/Linux

[matthewcroughan@aarch64:~]$ NIXPKGS_ALLOW_UNFREE=1 nix build github:MatthewCroughan/nixpkgs/mc/rkbin-rkboot-loaders#rkboot --impure

[matthewcroughan@aarch64:~]$ ls result/bin/
px30_loader_slc_v2.09.137.bin       rk3036_loader_v1.09.257.bin         rk3188_loader_v2.00.200.bin              rk3368_loader_v2.06.268.bin           rk3588_spl_loader_v1.16.113.bin
px30_loader_v2.09.135.bin           RK3066Loader_miniall.bin            rk322xh_loader_v1.19.250.bin             rk3399_loader_spinor_v1.30.114.bin    rknpu_lion_loader_v1.04.103.bin
px30_loader_wo_ftl_v2.09.135.bin    RK3066Loader_mini.bin               rk322x_loader_at_v1.10.256.bin           rk3399_loader_v1.30.130.bin           rv1106_download_v1.15.108.bin
px3se_loader_slc_v2.09.263.bin      RK3066Loader_uboot.bin              rk322x_loader_v1.10.256.bin              rk3399pro_loader_v1.30.126.bin        rv1106_idblock_v1.15.100.img
px3se_loader_v2.09.252.bin          rk3126_bootloader_v2.09.bin         rk3288_loader_v1.10.263.bin              rk3528_spl_loader_v1.09.105.bin       rv1106_idblock_v1.15.102.img
px5_loader_v2.06.258.bin            rk3126_loader_slc_v2.09.263.bin     rk3288_ubootloader_v1.11.bin             rk3562_spl_loader_v1.05.106.bin       rv110x_loader_v1.12.126.bin
px5_loader_v2.06.262.bin            rk3126_loader_v2.09.263.bin         rk3308_loader_uart4_v2.07.139.bin        rk3566_spl_loader_nand_v1.21.107.bin  rv1126_download_v1.09.109.bin
rk1808_loader_v1.06.105.bin         rk3126_loader_wo_ftl_v2.09.263.bin  rk3308_loader_v2.07.132.bin              rk3568_loader_spinand_v1.21.115.bin   rv1126_idblock_v1.09.110.img
rk1808_loader_wo_ftl_v1.06.106.bin  rk3128_bootloader_v2.12.bin         rk3308_loader_v2.07.139.bin              rk3568_pcie_loader_v2.10.bin          rv1126_loader_v1.09.123.bin
RK2738SPIBoot.bin                   rk3128_loader_slc_v2.12.265.bin     rk3308_loader_wo_ftl_v2.07.141.bin       rk356x_ramboot_loader_v1.21.108.bin   rv1126_ramboot_loader_v1.09.112.bin
RK3026Loader_miniall.bin            rk3128_loader_v2.12.263.bin         rk3326_loader_aarch32_slc_v2.09.136.bin  rk356x_spl_loader_nand_v1.21.107.bin  rv1126_spl_loader_v1.08.100.bin
RK3026Loader_mini.bin               rk3128x_loader_v1.08.257.bin        rk3326_loader_aarch32_v2.09.136.bin      rk356x_spl_loader_v1.20.113.bin       rv1126_spl_loader_v1.08.112.bin
RK3026Loader_uboot.bin              RK3168Loader_miniall.bin            rk3326_loader_slc_v2.09.137.bin          rk356x_spl_loader_v1.21.113.bin       rv1126_spl_loader_v1.09.101.bin
rk3032_loader_slc_v1.09.263.bin     RK3168Loader_mini.bin               rk3326_loader_v2.09.135.bin              rk3583_spl_loader_v1.14.113.bin       rv1126_spl_loader_v1.09.110.bin
rk3032_loader_v1.09.260.bin         RK3168Loader_uboot.bin              rk3328_loader_v1.19.250.bin              rk3588_download_v1.16.113.bin         SDBoot.bin
rk3036_bootloader_v1.09.bin         RK3188Loader_miniall.bin            rk3366_loader_v1.00.102.bin              rk3588_idblock_v1.16.113.img
rk3036_echo_loader_v1.09.236.bin    RK3188Loader_mini.bin               rk3368_bootloader_v2.06.bin              rk3588_pcie_loader_v2.10.bin
rk3036_loader_slc_v1.09.263.bin     RK3188Loader_uboot.bin              rk3368h_loader_v2.06.268.bin             rk3588_ramboot_loader_v1.16.106.bin
```

Reply:
No, this was run on an M2 macbook running NixOS.

Reply:
Ah yeah, this here smells like an asahi-specific issue (spawning @yu-re-ka ):

> Error: PageSize configuration is wrong: configured with 4096, but got 16384

Reply:
No matter what, the derivation should not succeed if there's zero files produced.

Can you add a check for that, or explicitly skip the ones failing on your machine?

Reply:
@flokli I've added a check. There are none that fail on specific machines, it's just that given a revision of the repository, not all outputs will compile, which is valuable to keep track of over time.